design/design-doc-agent.md

```diff
@@ -0,0 +1,730 @@
+remote cluster into `bar` namespace of the central cluster instead of the
+default target namespace which is `ns-in-central` in this example. But it will
+keep using the default credentials `Secret`. In case you don't want different
+namespaces to share the same credentials `Secret`, then you can override this
+setting, too:
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: foo
+  annotations:
+    "agent.crossplane.io/target-namespace": bar
+    "agent.crossplane.io/credentials-secret-name": my-other-sa-creds
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-sa-creds
+  namespace: foo
+type: Opaque
+data:
+  kubeconfig: MWYyZDFlMmU2N2Rm...
+```
+
+Now, all the requirements in the `foo` namespace of the remote cluster will be
+synced to `bar` namespace of the central cluster and all of the operations will
+be done using the credentials in `my-other-sa-creds`.
+
+There is also the option to automate the namespace pairings in a way that if
+requirement `A` is deployed in namespace `foo` of the remote cluster, then it
+gets synced to namespace `foo` of the central cluster. You can enable this
+automation using a flag. Helm command would look like:
+```bash
+helm install crossplane/agent \
+  --set default-credentials-secret-name=my-sa-creds \
+  --set match-namespaces=true
+```
+
+As with all cases, you can override specific pairings via annotations on the
+namespaces of the remote cluster.
+
+#### Conflicts
+
+The agent does not allow any conflicts to happen, meaning if two requirements
+created with the same name by different agents in some namespace in the pool of
+clusters, then it will reject syncing it to the central cluster. In order to do
+that, there needs to be unique identifier of the source of the requirement.
+
+Let's go over the different setups and consider conflict cases:
+* Each namespace is annotated with the target namespace.
+  * No conflict for single remote cluster.
+  * Could conflict if more than one remote cluster is connected and its namespaces have
+    common annotations with other namespaces of other remote clusters.
+* Each namespace targets a namespace with the same name in the central cluster.
+  * No conflict for single remote cluster.
+  * Could conflict if more than one cluster is connected and its namespaces have
+    the same names.
+* All namespaces roll up to a default namespace.
+  * Could conflict for single remote cluster if different namespaces have
+    requirements with same name.
+  * Could conflict if more than one cluster is connected to the same namespace
+    but it's less likely to be in that situation since the default namespace
+    selection is done explicitly during agent setup, probably by an admin.
+
+In order to prevent conflicts, the agent will add two annotations to the
+requirements it syncs:
+* `agent.crossplane.io/source-namespace`: The name of the namespace in the
+  remote cluster.
+* `agent.crossplane.io/source-uid`: The unique identifier of the remote
```

Comment:
`source-cluster-uid`? `source-uid` sounds like it could mean the UID of the remote requirement.